### ISSUE TITLE
On roadmap of label augmentation

### ISSUE THREAD

**Issue by crucifyml** [NVIDIA/DALI#219]
TL;DR

This is an issue with strong subjective opinion.
Thanks for reading and I apologize if offend you.

Do not mess up DALI with label modification too much, such as complex bounding box modification in detection task.

Implementing Reader for common task is ok, which reduces the difficulty of getting started. However, there are so may different algorithms for different tasks such as detection, classification, human pose estimation, face alignment, face verification and etc. These algorithms may require different labels in different forms.

It's **impossible** to implement a reader for each algorithm.

So, DALI should focus on providing a blazing fast image reader and a bunch of augmentation operators for image/mask.
To achieve the purpose of augmenting labels along with the image. I came up with 2 ideas.

**1. Outside graph** #213 
We can provide description vectors for every augmentation operation as computation graph's option output. For example, make Rotation angles, WarpAffine matrix available in outputs. Algorithm developer can augment labels after got the original label and according augmentation descriptions.

**2. Inside graph**
API for implement operation in python. So algorithm developer can augment label in python.
Also two ways to augment labels with image.
1. Pre-augment python operator: Python operator take label and image shape as input, output augmented labels and augmentation arguments. Augment image using pre-calculated augmentation arguments.
2. Post-augment python operator: Python operator take support operator and label as input, output augmented labels. Augment image using the same support operator as label's.

To sum up, we should make augmentation description/arguments available to algorithm developer. The augmentation of label should be exposed to algorithm developer. If the augmentation arguments can be manually set by algorithm developer (2.1 above), it would be most flexible. For PyTorch, the implementation would be very easy. For TensorFlow, static computation graph makes the work harder.

Discussion welcomed.

**Comment by JanuszL** [comment]
Hi,
Yes, that is true, it is hard to cover all uses cases. Now we are focusing on those easiest one where user can benefit most in the same time. In the long term we want to allow user to create own operators in Python - rather long term goal for now.
Keep in mind that labels and images augmentations cannot be easily decoupled - sometimes, like in case of SSD we check if random crop is valid for the bbox first and then we do cropping on the corresponding image - #196.
1) Idea sounds good, 2) is PythonOp we want to support someday.
Internally tracked as DALI-287.

**Comment by JanuszL** [comment]
Hi,

You can check [this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/math/geometric_transforms.html) that shows how to manipulate Keypoints. I think that should fulfill your request.
